conjure-error/src/error.rs

```diff
@@ -0,0 +1,380 @@
+
+impl ThrottleError {
+    /// Returns the amount of time the client should wait before retrying, if provided.
+    #[inline]
+    pub fn duration(&self) -> Option<Duration> {
+        self.duration
+    }
+}
+
+/// Information about an unavailable error.
+#[derive(Debug)]
+pub struct UnavailableError(());
+
+/// Information about the specific type of an `Error`.
+#[derive(Debug)]
+pub enum ErrorKind {
+    /// A general service error.
+    Service(SerializableError),
+    /// A QoS error indicating that the client should throttle itself.
+    Throttle(ThrottleError),
+    /// A QoS error indicating that the server was unable to handle the request.
+    Unavailable(UnavailableError),
+    #[doc(hidden)]
+    __NonExhaustive,
+}
+
+#[derive(Debug)]
+struct Inner {
+    cause: Box<dyn error::Error + Sync + Send>,
+    cause_safe: bool,
+    kind: ErrorKind,
+    safe_params: HashMap<Cow<'static, str>, Value>,
+    unsafe_params: HashMap<Cow<'static, str>, Value>,
+    backtraces: Vec<Backtrace>,
+}
+
+/// A standard error type for network services.
+///
+/// An error consists of several components:
+///
+/// * The cause of the error, represented as a type implementing the Rust `Error` trait. The cause can either be
+///     declared safe or unsafe to log.
+/// * The error's kind, indicating how the service should handle the error e.g. in a response to a client.
+/// * Backtraces, including one taken at the time the error was created.
+/// * Parameters adding extra context about the error. They can be declared either safe or unsafe to log.
+///
+/// Note that this type does *not* implement the standard library's `Error` trait.
+#[derive(Debug)]
+pub struct Error(Box<Inner>);
+
+impl Error {
+    /// Creates a service error with an unsafe cause.
+    pub fn service<E, T>(cause: E, error_type: T) -> Error
+    where
+        E: Into<Box<dyn error::Error + Sync + Send>>,
+        T: ErrorType + Serialize,
+    {
+        Error::service_inner(
+            cause.into(),
+            false,
+            crate::encode(&error_type),
+            error_type.safe_args(),
+        )
+    }
+
+    /// Creates a service error with a safe cause.
+    pub fn service_safe<E, T>(cause: E, error_type: T) -> Error
+    where
+        E: Into<Box<dyn error::Error + Sync + Send>>,
+        T: ErrorType + Serialize,
+    {
+        Error::service_inner(
+            cause.into(),
+            true,
+            crate::encode(&error_type),
+            error_type.safe_args(),
+        )
+    }
+
+    fn service_inner(
+        cause: Box<dyn error::Error + Sync + Send>,
+        cause_safe: bool,
+        error: SerializableError,
+        safe_args: &[&str],
+    ) -> Error {
+        let mut safe_params = HashMap::new();
+        let mut unsafe_params = HashMap::new();
+
+        for (key, value) in error.parameters() {
+            let key = Cow::Owned(key.clone());
+            let value = Value::String(value.clone());
+            if safe_args.contains(&&*key) {
```

Comment:
Can this be simplified to `&key`?

Reply:
Ah, we're relying on `Cow`'s implementation of `Deref`.

Reply:
No - we need to go from `&Cow<str>` to `&str`.